### PR TITLE
ORC-1482: Adaptation to read ORC files created by CUDF

### PR DESCRIPTION
**Pull request by guiyanakuang**
### What changes were proposed in this pull request?

This pr is aimed at adapting to read ORC files created by CUDF, which may have missing statistics in their DOUBLE/FLOAT columns.

### Why are the changes needed?

Official ORC readers can't read CUDF-created ORC files properly. 

### How was this patch tested?

Added UT.

### CODE REVIEW

**Comment by guiyanakuang** [comment]
Marked as a draft for now, with additional unit tests later on

**Comment by guiyanakuang** [comment]
@jlowe Can you help verify if this pr fixes the bug?

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java
Comment (dongjoon-hyun) [inline java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:1252]: I'm curious why we need to change this. According to the function description, this is used in the test code already, isn't it?
> Get the count of exceptions for testing.

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java
Comment (dongjoon-hyun) [inline java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:699]: This doesn't match with the following goal of PR.
> This pr is aimed at adapting to read ORC files created by CUDF, which may have missing statistics in their DOUBLE/FLOAT columns.

Previously, we assume that FLOAT type also has `DoubleColumnStatistics`. Now, it breaks the assumption in case of `FLOAT` type with `FloatColumnStatistics`.

In order to match with the PR description, do we need a null check here?

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java
Comment (guiyanakuang) [reply java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:1252]: Unit test that uses this method, but the test file is not in the same package as RecordReaderImpl.java

File: java/core/src/test/org/apache/orc/TestColumnStatistics.java
Comment (dongjoon-hyun) [inline java/core/src/test/org/apache/orc/TestColumnStatistics.java:707]: Shall we move `testDoubleColumnWithoutDoubleStatistics` to `TestRecordReaderImpl`?

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java
Comment (guiyanakuang) [reply java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:699]: When specific statistics are missing, cs is a `ColumnStatistics` object.

File: java/core/src/test/org/apache/orc/TestColumnStatistics.java
Comment (guiyanakuang) [reply java/core/src/test/org/apache/orc/TestColumnStatistics.java:707]: Let's refactor this pr later

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java
Comment (dongjoon-hyun) [reply java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:699]: Oh, got it.

**Review by jlowe — APPROVED** [review]
> @jlowe Can you help verify if this pr fixes the bug?

I built a new version of the ORC 1.8.4 jars with this patch, dropped them into Spark 3.4.1 and verified that I no longer get the exception when trying to do predicate pushdown on a cudf-generated ORC file that happens to have no double stats.  Thanks for the quick turnaround on this!

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM if CI passes. Thank you, @guiyanakuang and @jlowe .

**Comment by dongjoon-hyun** [comment]
cc @wgtmac 

Do you want to have this at your release?

**Comment by wgtmac** [comment]
Yes, I think this should be useful. @dongjoon-hyun 

@guiyanakuang Please backport it to branch-1.8 once merged. Thanks!

**Comment by dongjoon-hyun** [comment]
Thank you so much, @wgtmac . Then, let's target #1594 to your release too.

**Comment by dongjoon-hyun** [comment]
Since #1598 exists for 1.8.5, let me switch this to 1.9.2.

**Pull request closed without merging**
